### PR TITLE
Use Roslyn support for RuntimeHelpers.CreateSpan (or field caching downlevel)

### PR DESCRIPTION
**Pull request by stephentoub**

### CODE REVIEW

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/InflaterManaged.cs [hunk 1]
```diff
@@ -18,15 +18,15 @@ internal sealed class InflaterManaged
...
+        private static ReadOnlySpan<int> LengthBase => new int[]
         {
             3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 15, 17, 19, 23, 27, 31, 35, 43, 51,
             59, 67, 83, 99, 115, 131, 163, 195, 227, 3
         };
 
         // The base distance for distance code 0 - 31
         // The real distance for a distance code is  distanceBasePosition[code] + (value stored in extraBits)
-        private static readonly int[] s_distanceBasePosition =
+        private static ReadOnlySpan<int> DistanceBasePosition => new int[]
```
Comment (teo-tsirpanis) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/InflaterManaged.cs:29]: Can these be `byte` and `ushort` respectively?

And in general would it be worth it to use the smallest possible integer type for size savings?

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs [hunk 2]
```diff
@@ -549,7 +549,7 @@ private static bool TryCreateDateTime(DateTimeParseData parseData, DateTimeKind
             return true;
         }
 
-        private static readonly int[] s_daysToMonth365 = { 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365 };
-        private static readonly int[] s_daysToMonth366 = { 0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366 };
+        private static ReadOnlySpan<uint> DaysToMonth365 = { 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365 };
```
Comment (teo-tsirpanis) [inline src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs:552]:
```suggestion
        private static ReadOnlySpan<uint> DaysToMonth365 => { 0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365 };
```

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs [hunk 3]
```diff
@@ -549,7 +549,7 @@ private static bool TryCreateDateTime(DateTimeParseData parseData, DateTimeKind
...
+        private static ReadOnlySpan<uint> DaysToMonth366 = { 0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366 };
```
Comment (teo-tsirpanis) [inline src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs:553]:
```suggestion
        private static ReadOnlySpan<uint> DaysToMonth366 => { 0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366 };
```

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/InflaterManaged.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/InflaterManaged.cs]: They can be.

There is a size benefit to using a smaller type.  In general I haven't audited to make such changes, as that's not a new benefit; the data for these as arrays and as spans is stored almost exactly the same way.

**Comment by stephentoub** [comment]
Some compiler fixes in https://github.com/dotnet/roslyn/pull/65919.

**Comment by jcouv** [comment]
FYI, resolved the conflict

File: src/libraries/Common/src/System/Security/IdentityHelper.cs [hunk 4]
```diff
@@ -95,6 +87,7 @@ internal static string ToBase32StringSuitableForDirName(byte[] buff)
             // Create l chars using the last 5 bits of each byte.
             // Consume 3 MSB bits 5 bytes at a time.
 
+            const string Base32Chars = "abcdefghijklmnopqrstuvwxyz012345";
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Common/src/System/Security/IdentityHelper.cs:90]: > why not ReadOnlySpan to save also size?

What size difference are you referring to?

[the remark this answers is not in the document]
Comment (EgorBo) [reply src/libraries/Common/src/System/Security/IdentityHelper.cs:90]: My understanding that `ReadOnlySpan<char> Base32Chars => "string literal";` is not supported currently and there was a request in Roslyn for it (can't find it). So currently the only way is to do `=> new char[] { 'a', 'b'...` which is a bit uglier and probably not worth it?

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Common/src/System/Security/IdentityHelper.cs:90]: > ReadOnlySpan<char> Base32Chars => "string literal";

It's "supported", but there's nothing special about the codegen: it's just calling the implicit cast operator on the const string and returning that span.

> which is a bit uglier

"a bit"? :)

> and probably not worth it?

My question is really what is the "it" here?  What's the size difference cited and where does that come from?  I'm not seeing a difference in size whether I use `const string Blah = "abc"` or `static ReadOnlySpan<char> Blah => new char[] { 'a', 'b', 'c' }`

[the remark this answers is not in the document]
Comment (EgorBo) [reply src/libraries/Common/src/System/Security/IdentityHelper.cs:90]: > What's the size difference cited and where does that come from? I'm not seeing a difference in size whether I use const string Blah = "abc" or static ReadOnlySpan<char> Blah => new char[] { 'a', 'b', 'c' }

the string literal will be allocated on heap (in jit time) while it also occupies memory in the metadata, right? 
While in the RVA case nothing will be allocated on the heap and we'll be accessing data directly 🙂

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Common/src/System/Security/IdentityHelper.cs:90]: I thought Marek's comment was about IL size.  If this is instead referring to working set, then yeah, though for a string this small I don't think it's worth the increase code complexity going from:
```C#
const string Base32Chars = "abcdefghijklmnopqrstuvwxyz012345";
```
to
```C#
ReadOnlySpan<char> base32Chars => new char[]
{
    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
    '0', '1', '2', '3', '4', '5'
};
```
This particular file is also compiled into multiple projects, including at least one that targets downlevel, where this won't be an RVA and will instead allocate an array and store it into a static readonly field.

If others disagree, though, I can change it.

[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/Common/src/System/Security/IdentityHelper.cs:90]: > the string literal will be allocated on heap (in jit time) while it also occupies memory in the metadata, right

Yeah, my comment was that in this case the string version is bigger everywhere (disc, memory, code) and char[] version does not look that bad to me.

>  including at least one that targets downlevel

You could just ifdef 
```c#
#if NET8
ReadOnlySpan<char> base32Chars => new char[]
#else
private static readonly char[] base32Char =
#endif
```

Just suggestion ;-)

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Common/src/System/Security/IdentityHelper.cs:90]: > Yeah, my comment was that in this case the string version is bigger everywhere (disc, memory, code)

Can you elaborate on the disk and code part?

[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/Common/src/System/Security/IdentityHelper.cs:90]: > Can you elaborate on the disk and code part?

* Strings are UTF16 and ROS<char> are not
* For most codegens it's easier to optimize ROS index than string index nowadays

the difference will be obviously small so it's just general answer or am I wrong on these ?

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Common/src/System/Security/IdentityHelper.cs:90]: > Strings are UTF16 and ROS are not

`ReadOnlySpan<char>` is.  They're the same size, hence my pushing on it.

That said, for this particular case, I can make it `ReadOnlySpan<byte>` and use u8 for conciseness, so I'll do that.

**Comment by EgorBo** [comment]
Just curious - is it blocked by some other roslyn/illink bug currently? Can't wait to see it merged 🙂

**Comment by stephentoub** [comment]
> is it blocked by some other roslyn/illink bug currently?

Yes, it's blocked by a cecil / illink bug that's resulting in some fields being misaligned.

**Comment by marek-safar** [comment]
@stephentoub #79449 is the last blocker

**Comment by stephentoub** [comment]
> https://github.com/dotnet/runtime/pull/79449 is the last blocker

Excellent, thanks.

**Comment by stephentoub** [comment]
This is ready for review.

**Review by EgorBo — APPROVED** [review]
It seems that `DaysToMonth365` and `DaysToMonth366` are duplicated in several files but I assume it's not a big deal

**Comment by stephentoub** [comment]
> It seems that DaysToMonth365 and DaysToMonth366 are duplicated in several files but I assume it's not a big deal

This PR isn't adding any duplication.  It removes some of the duplication already there, and then what's left remains duplicated because they're either defined in different assemblies or using different types.  We could obviously reduce that source duplication by putting them in a shared file, but I chose not to do so as part of this PR.

**Comment by stephentoub** [comment]
@sbomer, this is still hitting an alignment issue, on both x86 and arm64 legs 😦 
```
      System.ArgumentException : Value does not fall within the expected range.
      Stack Trace:
           at System.Runtime.CompilerServices.RuntimeHelpers.GetSpanDataFrom(RuntimeFieldHandle fldHandle, RuntimeTypeHandle targetTypeHandle, Int32& count)
        /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs(100,0): at System.Runtime.CompilerServices.RuntimeHelpers.CreateSpan[T](RuntimeFieldHandle fldHandle)
        /_/src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs(140,0): at System.Decimal.DecCalc.get_DoublePowers10()
        /_/src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs(1732,0): at System.Decimal.DecCalc.VarDecFromR8(Double input, DecCalc& result)
        /_/src/libraries/System.Private.CoreLib/src/System/Decimal.cs(195,0): at System.Decimal..ctor(Double value)
        /_/src/libraries/System.Private.CoreLib/src/System/Decimal.cs(935,0): at System.Decimal.op_Explicit(Double value)
    Microsoft.Extensions.Configuration.FileExtensions.Test.FileConfigurationBuilderExtensionsTest.GetFileProvider_ReturnPhysica
```
Is this somehow not picking up your fixes?  Are there more issues lurking?

**Comment by MichalStrehovsky** [comment]
> Is this somehow not picking up your fixes? Are there more issues lurking?

#79449 hasn't merged yet.

**Comment by stephentoub** [comment]
> https://github.com/dotnet/runtime/pull/79449 hasn't merged yet

Thanks. 🤦‍♂️  Somehow I saw the link as the larger dependency update that did merge; too many tabs open I guess. Will try again now that it has merged.

**Pull request merged**